fili-core/src/main/java/com/yahoo/bard/webservice/web/handlers/EtagCacheRequestHandler.java

```diff
@@ -0,0 +1,174 @@
+import com.yahoo.bard.webservice.metadata.QuerySigningService;
+import com.yahoo.bard.webservice.util.Utils;
+import com.yahoo.bard.webservice.web.DataApiRequest;
+import com.yahoo.bard.webservice.web.responseprocessors.DruidJsonResponseContentKeys;
+import com.yahoo.bard.webservice.web.responseprocessors.EtagCacheResponseProcessor;
+import com.yahoo.bard.webservice.web.responseprocessors.LoggingContext;
+import com.yahoo.bard.webservice.web.responseprocessors.ResponseProcessor;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * Request handler to check the cache for a matching eTag and either return the cached result or send the next
+ * handler.
+ * <p>
+ * It also wraps the response processor so that valid, non-partial JSON responses are cached
+ */
+public class EtagCacheRequestHandler extends BaseDataRequestHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EtagCacheRequestHandler.class);
+    private static final MetricRegistry REGISTRY = MetricRegistryFactory.getRegistry();
+    public static final Meter CACHE_HITS = REGISTRY.meter("queries.meter.cache.hits");
+    public static final Meter CACHE_POTENTIAL_HITS = REGISTRY.meter("queries.meter.cache.potential_hits");
+    public static final Meter CACHE_MISSES = REGISTRY.meter("queries.meter.cache.misses");
+    public static final Meter CACHE_REQUESTS = REGISTRY.meter("queries.meter.cache.total");
+
+    protected final @NotNull DataRequestHandler next;
+    protected final @NotNull TupleDataCache<String, Long, String> dataCache;
+    protected final @NotNull QuerySigningService<Long> querySigningService;
+
+    /**
+     * Build a Cache request handler.
+     *
+     * @param next  The next handler in the chain
+     * @param dataCache  The cache instance
+     * @param querySigningService  The service to generate query signatures
+     * @param mapper  The mapper for all JSON processing
+     */
+    @SuppressWarnings("unchecked")
+    public EtagCacheRequestHandler(
+            DataRequestHandler next,
+            @NotNull DataCache<?> dataCache,
+            QuerySigningService<?> querySigningService,
+            ObjectMapper mapper
+    ) {
+        super(mapper);
+        this.next = next;
+        this.dataCache = (TupleDataCache<String, Long, String>) dataCache;
+        this.querySigningService = (QuerySigningService<Long>) querySigningService;
+    }
+
+    @Override
+    public boolean handleRequest(
+            final RequestContext context,
+            final DataApiRequest request,
+            final DruidAggregationQuery<?> druidQuery,
+            final ResponseProcessor response
```

Comment:
I modified and it's needed outside in constructing `EtagCacheResponseProcessor `